enhancements/big-bundle-unpacking.md

```diff
@@ -0,0 +1,138 @@
+---
+title: big-bundle-unpacking
+authors:
+  - "@njhale"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-08-12
+last-updated: 2020-09-01
+status: implementable
+see-also: - "https://docs.google.com/document/d/1WWbZCyG1sw3vi-1gtMOshmaTaYiMht3cb-Up4NAx4_I/edit#"
+
+---
+
+# Big Bundle Unpacking
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+To ensure Operator Bundles of arbitrary size can be unpacked and used in a Kubernetes cluster by OLM, bundle content will be partitioned across several ConfigMaps. This unpacking behavior will be abstracted behind a new Kubernetes API to allow for future extensibility.
+
+## Motivation
+
+Currently, OLM stages Operator Bundles within a cluster before installation by writing their content into ConfigMaps. There is a 1:1 relationship between bundle and ConfigMap; i.e. each bundle is unpacked to its own ConfigMap. A limitation of this staging method is that the maximum size of a ConfigMap for a given k8s cluster becomes the upper bound for the size of a bundle installed to it.
+
+### Goals
+
+- Enable OLM to install operators from bundles containing more content than fits in a single ConfigMap
+
+### Non-Goals
+
+- Increase the upper storage bound for API resources in general
+- Support bundles containing individual resources that surpass the upper storage bound of a cluster
+
+## Proposal
+
+A new Kubernetes resource kind will be introduced into the `porcelain.operators.coreos.com/v1` group version: `Bundle`. A `Bundle` resource will abstract the pulling and unpacking of an operator bundle from clients (OLM and cluster admins) and will feature the following fields:
+
+| ----- | ----------- |
+| spec.from | A union type which locates the bundle (e.g. a docker image digest) |
+| status.conditions | [Standard k8s conditions](https://github.com/kubernetes/apimachinery/blob/94222d04a59075a01fddedd696037db9e61db6e9/pkg/apis/meta/v1/types.go#L1367), surfacing info about the unpack job |
+| status.mediaType | Discriminates the type of bundle pulled (e.g. `registry+v1`) |
+
+```yaml
+apiVersion: porcelain.operators.coreos.com/v1alpha1
+kind: Bundle
+metadata:
+  name: my-operator
+  namespace: olm
+spec:
+  from:
+    type: Image
+    ref: quay.io/njhale/my-operator@sha256:123def...
+status:
+  mediaType: registry+v1
+  conditions:
+    type: UnpackFailed
+    status: True
+    reason: InvalidBundle
+    message: "no manifests detected in bundle"
+```
+
+Most importantly, a `Bundle` will expose a `/content` subresource which -- upon a successful unpackaging -- will return a [JSON encoded representation]() of that bundle.
```

Comment:
Would it make sense to have the `/content` subresource accept a filename parameter and return only that specific manifest? If I understand correctly, the catalog-operator consume manifests step by step, so this can avoid some wasteful traffic between the catalog-operator and the aggregated api server serving that subresource.